modules/doctor/heads.go

```diff
@@ -0,0 +1,97 @@
+// Copyright 2020 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package doctor
+
+import (
+	"context"
+
+	repo_model "code.gitea.io/gitea/models/repo"
+	"code.gitea.io/gitea/modules/git"
+	"code.gitea.io/gitea/modules/log"
+)
+
+func synchronizeRepoHeads(ctx context.Context, logger log.Logger, autofix bool) error {
+	numRepos := 0
+	numHeadsBroken := 0
+	numDefaultBranchesBroken := 0
+	numReposUpdated := 0
+	err := iterateRepositories(ctx, func(repo *repo_model.Repository) error {
+		numRepos++
+		runOpts := &git.RunOpts{Dir: repo.RepoPath()}
+
+		head, _, headErr := git.NewCommand(ctx, "rev-parse", "HEAD").RunStdString(runOpts)
+		defaultBranch, _, defaultBranchErr := git.NewCommand(ctx, "rev-parse", repo.DefaultBranch).RunStdString(runOpts)
+
+		// what we expect: both HEAD and default branch point to the same commit
+		if headErr == nil && defaultBranchErr == nil && head == defaultBranch {
+			return nil
+		}
+
+		if headErr != nil {
+			numHeadsBroken++
+		}
+		if defaultBranchErr != nil {
+			numDefaultBranchesBroken++
+		}
+
+		// absolute failure: both HEAD and default branch point to invalid commits
+		if headErr != nil && defaultBranchErr != nil {
+			logger.Critical("Neither HEAD nor the default branch for %s/%s point to a valid commit", repo.OwnerName, repo.Name)
+			return nil
+		}
+
+		// if default branch is broken, let the user fix that in the UI
+		if defaultBranchErr != nil {
+			logger.Warn("Default branch for %s/%s doesn't point to a valid commit", repo.OwnerName, repo.Name)
+			return nil
+		}
+
+		// if we're not autofixing, that's all we can do
+		if !autofix {
+			return nil
+		}
+
+		// otherwise, let's try fixing HEAD
+		err := git.NewCommand(ctx, "switch", repo.DefaultBranch).Run(runOpts)
```

Comment:
Shouldn't this be `git symbolic-ref HEAD repo.DefaultBranch`?

Our repos should be bare so switch may not do the correct thing

Reply:
Thank you for finding this for me; I had trouble locating the right command.